File: packages/shared/src/contract.ts
/**
 * Values shared by the pipeline, the Worker and the web app, read from data/contract.json (the
 * Python pipeline reads the same file).
 */
import contract from "../../../data/contract.json";

/**
 * Narrow a list from the JSON file to the literal type the code is written against, failing
 * loudly (at build time and in tests) if the two drift apart.
 */
function literals<const T extends readonly string[]>(actual: readonly string[], expected: T): T {
  if (actual.length !== expected.length || actual.some((value, i) => value !== expected[i])) {
    throw new Error(`data/contract.json: expected [${expected}], found [${actual}]`);
  }
  return expected;
}

export const LOCALES = literals(contract.locales, ["it", "en"]);
export type Locale = (typeof LOCALES)[number];
export const DEFAULT_LOCALE: Locale = "it";

export function isLocale(value: unknown): value is Locale {
  return (LOCALES as readonly unknown[]).includes(value);
}

/** Lifetime of each kind of signed link, in seconds; 0 means the link never expires. */
export const TOKEN_TTL_SECONDS = contract.tokenTtlSeconds;
export type TokenPurpose = keyof typeof TOKEN_TTL_SECONDS;
export const TOKEN_PURPOSES = Object.keys(TOKEN_TTL_SECONDS) as TokenPurpose[];

/** Unconfirmed addresses are deleted after this many days (see the pipeline's daily purge). */
export const PENDING_RETENTION_DAYS = contract.pendingRetentionDays;

export type SitePage = keyof typeof contract.sitePaths;

/** Path of a page of the web app, relative to its base URL, e.g. `/it/manage/`. */
export function sitePath(page: SitePage, locale: Locale): string {
  return contract.sitePaths[page].replace("{locale}", locale);
}

/** Paths served by the Worker that email links point to. */
export const API_PATHS = contract.apiPaths;

/** Query and fragment parameters of the links sent by email. */
export const LINK_PARAMS = contract.linkParams;

/** Where the emails send people who need help: their replies are not read. */
export const ISSUES_URL = contract.issuesUrl;

export const INSTITUTION_TYPES = literals(contract.institutionTypes, [
  "university",
  "online_university",
  "research_institute",
  "afam",
]);
export type InstitutionType = (typeof INSTITUTION_TYPES)[number];

File: apps/pipeline/src/postthedoc/contract.py
"""Values shared with the Worker and the web app, read from data/contract.json.

The literal types below mirror the lists in the JSON file: loading a contract that does not
match them fails validation, so the two cannot drift apart silently.
"""

from pathlib import Path
from typing import Literal

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

Locale = Literal["it", "en"]
TokenPurpose = Literal["confirm", "manage", "unsubscribe"]
InstitutionType = Literal["university", "online_university", "research_institute", "afam"]
SitePage = Literal["home", "subscribe", "manage", "privacy", "support"]


class _Model(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, frozen=True, extra="forbid")


class TokenTtl(_Model):
    confirm: int
    manage: int
    unsubscribe: int  # 0: one-click unsubscribe links in old digests keep working

    def seconds(self, purpose: TokenPurpose) -> int:
        match purpose:
            case "confirm":
                return self.confirm
            case "manage":
                return self.manage
            case "unsubscribe":
                return self.unsubscribe


class SitePaths(_Model):
    home: str
    subscribe: str
    manage: str
    privacy: str
    support: str


class ApiPaths(_Model):
    confirm: str
    unsubscribe: str


class LinkParams(_Model):
    token: str
    welcome: str


class DailyEmails(_Model):
    provider: int  # the email provider's daily quota
    worker: int  # the Worker's share: confirmations and manage links

    @property
    def digests(self) -> int:
        """What is left for the daily digests, one per user."""
        return self.provider - self.worker


class Contract(_Model):
    model_config = ConfigDict(alias_generator=to_camel, frozen=True, extra="ignore")

    locales: tuple[Locale, ...]
    token_ttl_seconds: TokenTtl
    pending_retention_days: int
    daily_emails: DailyEmails
    site_paths: SitePaths
    api_paths: ApiPaths
    link_params: LinkParams
    issues_url: str
    institution_types: tuple[InstitutionType, ...]

    @classmethod
    def load(cls, path: Path) -> "Contract":
        return cls.model_validate_json(path.read_text(encoding="utf-8"))

    def site_path(self, page: SitePage, locale: Locale) -> str:
        """Path of a web app page relative to its base URL, e.g. "/it/manage/"."""
        path: str = getattr(self.site_paths, page)
        return path.replace("{locale}", locale)
